In [102]:
import pandas as pd

# Load dataset
df = pd.read_csv("data/churn_dataset.csv")  # Use relative path

# Display info
df.info()
df.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [103]:
df=df.iloc[:,1:]

In [104]:
df.duplicated().sum()

22

In [105]:
df.TotalCharges = pd.to_numeric(df.TotalCharges, errors='coerce')

In [106]:
print(df["TotalCharges"].dtype)

float64


In [107]:
df.isnull().sum()

gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [108]:
miss_column = df[df["TotalCharges"].isnull()]
print(miss_column)

      gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
488   Female              0     Yes        Yes       0           No   
753     Male              0      No        Yes       0          Yes   
936   Female              0     Yes        Yes       0          Yes   
1082    Male              0     Yes        Yes       0          Yes   
1340  Female              0     Yes        Yes       0           No   
3331    Male              0     Yes        Yes       0          Yes   
3826    Male              0     Yes        Yes       0          Yes   
4380  Female              0     Yes        Yes       0          Yes   
5218    Male              0     Yes        Yes       0          Yes   
6670  Female              0     Yes        Yes       0          Yes   
6754    Male              0      No        Yes       0          Yes   

         MultipleLines InternetService       OnlineSecurity  \
488   No phone service             DSL                  Yes   
753                 N

Nan values in TotalCharges column corresponds with 0 values at tenure it is because if there tenure is 0 there is no billing. So we change this Nan values with 0s.

In [109]:
df["TotalCharges"]=df["TotalCharges"].fillna(0)

In [110]:
df.isnull().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

We handled null values.
##Detecting anomalies

In [111]:
for col in df.select_dtypes(include=['object']):
    print(col, df[col].unique())

gender ['Female' 'Male']
Partner ['Yes' 'No']
Dependents ['No' 'Yes']
PhoneService ['No' 'Yes']
MultipleLines ['No phone service' 'No' 'Yes']
InternetService ['DSL' 'Fiber optic' 'No']
OnlineSecurity ['No' 'Yes' 'No internet service']
OnlineBackup ['Yes' 'No' 'No internet service']
DeviceProtection ['No' 'Yes' 'No internet service']
TechSupport ['No' 'Yes' 'No internet service']
StreamingTV ['No' 'Yes' 'No internet service']
StreamingMovies ['No' 'Yes' 'No internet service']
Contract ['Month-to-month' 'One year' 'Two year']
PaperlessBilling ['Yes' 'No']
PaymentMethod ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
Churn ['No' 'Yes']


It seems there is no anormally. For usefullnes we will convert boolean values to binary numeric values to use it in statistical calculations.

In [112]:
df["Partner"] = df["Partner"].map({"Yes": 1, "No": 0})
df["Dependents"] = df["Dependents"].map({"Yes": 1, "No": 0})
df["PhoneService"] = df["PhoneService"].map({"Yes": 1, "No": 0})
df["PaperlessBilling"] = df["PaperlessBilling"].map({"Yes": 1, "No": 0})
df["Churn"] = df["Churn"].map({"Yes": 1, "No": 0})


print(df[["Partner", "Dependents", "PhoneService", "PaperlessBilling", "Churn"]].head())

   Partner  Dependents  PhoneService  PaperlessBilling  Churn
0        1           0             0                 1      0
1        0           0             1                 0      0
2        0           0             1                 1      1
3        0           0             0                 0      0
4        0           0             1                 1      1



We have cleaned the data. 
Cleaning Summary
Our dataset consisted of a total of 7043 rows and 21 columns .
Our data is found in clean form .
Our dataset looks a bit tidy with no NaNs nor duplicated values.
TotalCharges column has null values that need to be handled, it has 11 nulls, and we replaced them by zero After finding that tenure values equal to zero.
No columns needed to be dropped.


Preprocessing will continue with column manipulation

In [113]:
df = df.map(lambda x: 1 if x == "Yes" else (0 if x == "No" else x))
df

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0     Female              0        1           0       1             0   
1       Male              0        0           0      34             1   
2       Male              0        0           0       2             1   
3       Male              0        0           0      45             0   
4     Female              0        0           0       2             1   
...      ...            ...      ...         ...     ...           ...   
7038    Male              0        1           1      24             1   
7039  Female              0        1           1      72             1   
7040  Female              0        1           1      11             0   
7041    Male              1        1           0       4             1   
7042    Male              0        0           0      66             1   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL              0            1   
1                    0             DSL              1            0   
2                    0             DSL              1            1   
3     No phone service             DSL              1            0   
4                    0     Fiber optic              0            0   
...                ...             ...            ...          ...   
7038                 1             DSL              1            0   
7039                 1     Fiber optic              0            1   
7040  No phone service             DSL              1            0   
7041                 1     Fiber optic              0            0   
7042                 0     Fiber optic              1            0   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                   0           0           0               0  Month-to-month   
1                   1           0           0               0        One year   
2                   0           0           0               0  Month-to-month   
3                   1           1           0               0        One year   
4                   0           0           0               0  Month-to-month   
...               ...         ...         ...             ...             ...   
7038                1           1           1               1        One year   
7039                1           0           1               1        One year   
7040                0           0           0               0  Month-to-month   
7041                0           0           0               0  Month-to-month   
7042                1           1           1               1        Two year   

      PaperlessBilling              PaymentMethod  MonthlyCharges  \
0                    1           Electronic check           29.85   
1                    0               Mailed check           56.95   
2                    1               Mailed check           53.85   
3                    0  Bank transfer (automatic)           42.30   
4                    1           Electronic check           70.70   
...                ...                        ...             ...   
7038                 1               Mailed check           84.80   
7039                 1    Credit card (automatic)          103.20   
7040                 1           Electronic check           29.60   
7041                 1               Mailed check           74.40   
7042                 1  Bank transfer (automatic)          105.65   

      TotalCharges  Churn  
0            29.85      0  
1          1889.50      0  
2           108.15      1  
3          1840.75      0  
4           151.65      1  
...            ...    ...  
7038       1990.50      0  
7039       7362.90      0  
7040        346.45      0  
7041        306.60      1  
7042       6844.50      0  

[7043 rows x 20 columns]

To change categorical variables into numerical values

In [114]:
df_dummies = pd.get_dummies(df).astype(float)
df_dummies.head()

SeniorCitizen  Partner  Dependents  tenure  PhoneService  PaperlessBilling  \
0            0.0      1.0         0.0     1.0           0.0               1.0   
1            0.0      0.0         0.0    34.0           1.0               0.0   
2            0.0      0.0         0.0     2.0           1.0               1.0   
3            0.0      0.0         0.0    45.0           0.0               0.0   
4            0.0      0.0         0.0     2.0           1.0               1.0   

   MonthlyCharges  TotalCharges  Churn  gender_Female  ...  StreamingMovies_0  \
0           29.85         29.85    0.0            1.0  ...                1.0   
1           56.95       1889.50    0.0            0.0  ...                1.0   
2           53.85        108.15    1.0            0.0  ...                1.0   
3           42.30       1840.75    0.0            0.0  ...                1.0   
4           70.70        151.65    1.0            1.0  ...                1.0   

   StreamingMovies_1  StreamingMovies_No internet service  \
0                0.0                                  0.0   
1                0.0                                  0.0   
2                0.0                                  0.0   
3                0.0                                  0.0   
4                0.0                                  0.0   

   Contract_Month-to-month  Contract_One year  Contract_Two year  \
0                      1.0                0.0                0.0   
1                      0.0                1.0                0.0   
2                      1.0                0.0                0.0   
3                      0.0                1.0                0.0   
4                      1.0                0.0                0.0   

   PaymentMethod_Bank transfer (automatic)  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      1.0   
4                                      0.0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                    0.0                             1.0   
1                                    0.0                             0.0   
2                                    0.0                             0.0   
3                                    0.0                             0.0   
4                                    0.0                             1.0   

   PaymentMethod_Mailed check  
0                         0.0  
1                         1.0  
2                         1.0  
3                         0.0  
4                         0.0  

[5 rows x 42 columns]